// @nuxt/content/dist/runtime/composables/utils.mjs
import { withBase } from "ufo";
import { useRuntimeConfig, useRequestEvent, useCookie, useRoute } from "#app";
import { unwrap, flatUnwrap } from "../markdown-parser/utils/node.mjs";
export const withContentBase = (url) => withBase(url, useRuntimeConfig().public.content.api.baseURL);
export const useUnwrap = () => ({
  unwrap,
  flatUnwrap
});
export const useContentDisabled = () => {
  console.warn("useContent is only accessible when you are using `documentDriven` mode.");
  console.warn("Learn more by visiting: https://content.nuxtjs.org/guide/writing/document-driven");
  throw new Error("useContent is only accessible when you are using `documentDriven` mode.");
};
export const navigationDisabled = () => {
  console.warn("Navigation is only accessible when you enable it in module options.");
  console.warn("Learn more by visiting: https://content.nuxtjs.org/api/configuration#navigation");
  throw new Error("Navigation is only accessible when you enable it in module options.");
};
export const addPrerenderPath = (path) => {
  const event = useRequestEvent();
  event.res.setHeader(
    "x-nitro-prerender",
    [
      event.res.getHeader("x-nitro-prerender"),
      path
    ].filter(Boolean).join(",")
  );
};
export const shouldUseClientDB = () => {
  const { experimental } = useRuntimeConfig().content;
  if (process.server) {
    return false;
  }
  if (experimental.clientDB) {
    return true;
  }
  const query = useRoute().query;
  if (Object.prototype.hasOwnProperty.call(query, "preview") && !query.preview) {
    return false;
  }
  if (query.preview || useCookie("previewToken").value) {
    if (process.dev) {
      console.warn("[content] Client DB enabled since a preview token is set (either in query or cookie).");
    }
    return true;
  }
  return false;
};
