// @nuxt/devtools/dist/runtime/plugins/devtools.client.mjs
import { createApp, markRaw } from "vue";
import { createHooks } from "hookable";
import { setupHooksDebug } from "../shared/hooks.mjs";
import Container from "./view/Container.vue";
import { closePanel, togglePanel } from "./view/state.mjs";
import { defineNuxtPlugin } from "#app";
import { useAppConfig } from "#imports";
export default defineNuxtPlugin((nuxt) => {
  if (typeof document === "undefined" || typeof window === "undefined")
    return;
  if (window.parent && window.self !== window.parent) {
    try {
      if (window.parent.__NUXT_DEVTOOLS_VIEW__ || window.parent.document.querySelector("#nuxt-devtools-container"))
        return;
    } catch (e) {
    }
  }
  const clientHooks = setupHooksDebug(nuxt.hooks);
  const client = markRaw({
    nuxt: markRaw(nuxt),
    appConfig: useAppConfig(),
    hooks: createHooks(),
    getClientHooksMetrics: () => Object.values(clientHooks),
    getClientPluginMetrics: () => {
      return globalThis.__NUXT_DEVTOOLS_PLUGINS_METRIC__ || [];
    },
    reloadPage() {
      location.reload();
    },
    closeDevTools: closePanel
  });
  const holder = document.createElement("div");
  holder.id = "nuxt-devtools-container";
  holder.setAttribute("data-v-inspector-ignore", "true");
  document.body.appendChild(holder);
  addEventListener("keypress", (e) => {
    if (e.code === "KeyD" && e.altKey && e.shiftKey)
      togglePanel();
  });
  const app = createApp(Container, { client });
  app.mount(holder);
});
