// @nuxt/devtools/dist/cli.mjs
import fs from 'node:fs';
import consola from 'consola';
import { execa } from 'execa';
import { readUser, writeUser } from 'rc9';
import require$$1 from 'path';
import require$$1$1 from 'os';
import require$$0__default from 'fs';
import { resolve } from 'pathe';
import { n as name } from './shared/devtools.696fb428.mjs';

var globalDirs = {};

const { hasOwnProperty } = Object.prototype;

const eol = typeof process !== 'undefined' &&
  process.platform === 'win32' ? '\r\n' : '\n';

const encode = (obj, opt) => {
  const children = [];
  let out = '';

  if (typeof opt === 'string') {
    opt = {
      section: opt,
      whitespace: false,
    };
  } else {
    opt = opt || Object.create(null);
    opt.whitespace = opt.whitespace === true;
  }

  const separator = opt.whitespace ? ' = ' : '=';

  for (const k of Object.keys(obj)) {
    const val = obj[k];
    if (val && Array.isArray(val)) {
      for (const item of val)
        out += safe(k + '[]') + separator + safe(item) + '\n';
    } else if (val && typeof val === 'object')
      children.push(k);
    else
      out += safe(k) + separator + safe(val) + eol;
  }

  if (opt.section && out.length)
    out = '[' + safe(opt.section) + ']' + eol + out;

  for (const k of children) {
    const nk = dotSplit(k).join('\\.');
    const section = (opt.section ? opt.section + '.' : '') + nk;
    const { whitespace } = opt;
    const child = encode(obj[k], {
      section,
      whitespace,
    });
    if (out.length && child.length)
      out += eol;

    out += child;
  }

  return out
};

const dotSplit = str =>
  str.replace(/\1/g, '\u0002LITERAL\\1LITERAL\u0002')
    .replace(/\\\./g, '\u0001')
    .split(/\./)
    .map(part =>
      part.replace(/\1/g, '\\.')
        .replace(/\2LITERAL\\1LITERAL\2/g, '\u0001'));

const decode = str => {
  const out = Object.create(null);
  let p = out;
  let section = null;
  //          section     |key      = value
  const re = /^\[([^\]]*)\]$|^([^=]+)(=(.*))?$/i;
  const lines = str.split(/[\r\n]+/g);

  for (const line of lines) {
    if (!line || line.match(/^\s*[;#]/))
      continue
    const match = line.match(re);
    if (!match)
      continue
    if (match[1] !== undefined) {
      section = unsafe(match[1]);
      if (section === '__proto__') {
        // not allowed
        // keep parsing the section, but don't attach it.
        p = Object.create(null);
        continue
      }
      p = out[section] = out[section] || Object.create(null);
      continue
    }
    const keyRaw = unsafe(match[2]);
    const isArray = keyRaw.length > 2 && keyRaw.slice(-2) === '[]';
    const key = isArray ? keyRaw.slice(0, -2) : keyRaw;
    if (key === '__proto__')
      continue
    const valueRaw = match[3] ? unsafe(match[4]) : true;
    const value = valueRaw === 'true' ||
      valueRaw === 'false' ||
      valueRaw === 'null' ? JSON.parse(valueRaw)
      : valueRaw;

    // Convert keys with '[]' suffix to an array
    if (isArray) {
      if (!hasOwnProperty.call(p, key))
        p[key] = [];
      else if (!Array.isArray(p[key]))
        p[key] = [p[key]];
    }

    // safeguard against resetting a previously defined
    // array by accidentally forgetting the brackets
    if (Array.isArray(p[key]))
      p[key].push(value);
    else
      p[key] = value;
  }

  // {a:{y:1},"a.b":{x:2}} --> {a:{y:1,b:{x:2}}}
  // use a filter to return the keys that have to be deleted.
  const remove = [];
  for (const k of Object.keys(out)) {
    if (!hasOwnProperty.call(out, k) ||
        typeof out[k] !== 'object' ||
        Array.isArray(out[k]))
      continue

    // see if the parent section is also an object.
    // if so, add it to that, and mark this one for deletion
    const parts = dotSplit(k);
    let p = out;
    const l = parts.pop();
    const nl = l.replace(/\\\./g, '.');
    for (const part of parts) {
      if (part === '__proto__')
        continue
      if (!hasOwnProperty.call(p, part) || typeof p[part] !== 'object')
        p[part] = Object.create(null);
      p = p[part];
    }
    if (p === out && nl === l)
      continue

    p[nl] = out[k];
    remove.push(k);
  }
  for (const del of remove)
    delete out[del];

  return out
};

const isQuoted = val =>
  (val.charAt(0) === '"' && val.slice(-1) === '"') ||
    (val.charAt(0) === "'" && val.slice(-1) === "'");

const safe = val =>
  (typeof val !== 'string' ||
    val.match(/[=\r\n]/) ||
    val.match(/^\[/) ||
    (val.length > 1 &&
     isQuoted(val)) ||
    val !== val.trim())
    ? JSON.stringify(val)
    : val.replace(/;/g, '\\;').replace(/#/g, '\\#');

const unsafe = (val, doUnesc) => {
  val = (val || '').trim();
  if (isQuoted(val)) {
    // remove the single quotes before calling JSON.parse
    if (val.charAt(0) === "'")
      val = val.substr(1, val.length - 2);

    try {
      val = JSON.parse(val);
    } catch (_) {}
  } else {
    // walk the val to find the first not-escaped ; character
    let esc = false;
    let unesc = '';
    for (let i = 0, l = val.length; i < l; i++) {
      const c = val.charAt(i);
      if (esc) {
        if ('\\;#'.indexOf(c) !== -1)
          unesc += c;
        else
          unesc += '\\' + c;

        esc = false;
      } else if (';#'.indexOf(c) !== -1)
        break
      else if (c === '\\')
        esc = true;
      else
        unesc += c;
    }
    if (esc)
      unesc += '\\';

    return unesc.trim()
  }
  return val
};

var ini = {
  parse: decode,
  decode,
  stringify: encode,
  encode,
  safe,
  unsafe,
};

(function (exports) {
	const path = require$$1;
	const os = require$$1$1;
	const fs = require$$0__default;
	const ini$1 = ini;

	const isWindows = process.platform === 'win32';

	const readRc = filePath => {
		try {
			return ini$1.parse(fs.readFileSync(filePath, 'utf8')).prefix;
		} catch {}
	};

	const getEnvNpmPrefix = () => {
		// TODO: Remove the `.reduce` call.
		// eslint-disable-next-line unicorn/no-array-reduce
		return Object.keys(process.env).reduce((prefix, name) => {
			return /^npm_config_prefix$/i.test(name) ? process.env[name] : prefix;
		}, undefined);
	};

	const getGlobalNpmrc = () => {
		if (isWindows && process.env.APPDATA) {
			// Hardcoded contents of `c:\Program Files\nodejs\node_modules\npm\npmrc`
			return path.join(process.env.APPDATA, '/npm/etc/npmrc');
		}

		// Homebrew special case: `$(brew --prefix)/lib/node_modules/npm/npmrc`
		if (process.execPath.includes('/Cellar/node')) {
			const homebrewPrefix = process.execPath.slice(0, process.execPath.indexOf('/Cellar/node'));
			return path.join(homebrewPrefix, '/lib/node_modules/npm/npmrc');
		}

		if (process.execPath.endsWith('/bin/node')) {
			const installDir = path.dirname(path.dirname(process.execPath));
			return path.join(installDir, '/etc/npmrc');
		}
	};

	const getDefaultNpmPrefix = () => {
		if (isWindows) {
			const {APPDATA} = process.env;
			// `c:\node\node.exe` → `prefix=c:\node\`
			return APPDATA ? path.join(APPDATA, 'npm') : path.dirname(process.execPath);
		}

		// `/usr/local/bin/node` → `prefix=/usr/local`
		return path.dirname(path.dirname(process.execPath));
	};

	const getNpmPrefix = () => {
		const envPrefix = getEnvNpmPrefix();
		if (envPrefix) {
			return envPrefix;
		}

		const homePrefix = readRc(path.join(os.homedir(), '.npmrc'));
		if (homePrefix) {
			return homePrefix;
		}

		if (process.env.PREFIX) {
			return process.env.PREFIX;
		}

		const globalPrefix = readRc(getGlobalNpmrc());
		if (globalPrefix) {
			return globalPrefix;
		}

		return getDefaultNpmPrefix();
	};

	const npmPrefix = path.resolve(getNpmPrefix());

	const getYarnWindowsDirectory = () => {
		if (isWindows && process.env.LOCALAPPDATA) {
			const dir = path.join(process.env.LOCALAPPDATA, 'Yarn');
			if (fs.existsSync(dir)) {
				return dir;
			}
		}

		return false;
	};

	const getYarnPrefix = () => {
		if (process.env.PREFIX) {
			return process.env.PREFIX;
		}

		const windowsPrefix = getYarnWindowsDirectory();
		if (windowsPrefix) {
			return windowsPrefix;
		}

		const configPrefix = path.join(os.homedir(), '.config/yarn');
		if (fs.existsSync(configPrefix)) {
			return configPrefix;
		}

		const homePrefix = path.join(os.homedir(), '.yarn-config');
		if (fs.existsSync(homePrefix)) {
			return homePrefix;
		}

		// Yarn supports the npm conventions but the inverse is not true
		return npmPrefix;
	};

	exports.npm = {};
	exports.npm.prefix = npmPrefix;
	exports.npm.packages = path.join(npmPrefix, isWindows ? 'node_modules' : 'lib/node_modules');
	exports.npm.binaries = isWindows ? npmPrefix : path.join(npmPrefix, 'bin');

	const yarnPrefix = path.resolve(getYarnPrefix());
	exports.yarn = {};
	exports.yarn.prefix = yarnPrefix;
	exports.yarn.packages = path.join(yarnPrefix, getYarnWindowsDirectory() ? 'Data/global/node_modules' : 'global/node_modules');
	exports.yarn.binaries = path.join(exports.yarn.packages, '.bin');
} (globalDirs));

const RC_PATH = ".nuxtrc";
function enable(rc, path) {
  const modulePath = resolve(globalDirs.npm.packages, name);
  const targetPath = resolve(modulePath, "module.cjs").replace(/\\/g, "/");
  if (!fs.existsSync(targetPath))
    throw new Error("Failed to locate the global Nuxt Devtools module. You may try it again");
  removeModule(rc);
  if (!rc.modules?.includes(targetPath))
    rc.modules = [...rc.modules || [], targetPath];
  if (!rc.devtoolsGlobal)
    rc.devtoolsGlobal = {};
  if (!rc.devtoolsGlobal.projects)
    rc.devtoolsGlobal.projects = [];
  if (!rc.devtoolsGlobal.projects.includes(path))
    rc.devtoolsGlobal.projects.push(path);
}
function disable(rc, path) {
  if (rc.devtoolsGlobal?.projects?.includes(path)) {
    rc.devtoolsGlobal.projects = rc.devtoolsGlobal.projects.filter((p) => p !== path);
    return true;
  }
  if (!rc.devtoolsGlobal.projects?.length)
    removeModule(rc);
  return false;
}
const pathRegex = /[\\\/]@nuxt[\\\/]devtools[\\\/]/;
function removeModule(rc) {
  if (Array.isArray(rc.modules))
    rc.modules = rc.modules.filter((p) => !p.match(pathRegex));
  return true;
}
async function run() {
  const args = process.argv.slice(2);
  const command = args[0];
  const cwd = process.cwd();
  const rc = readUser(RC_PATH);
  if (command === "enable") {
    consola.info("Installed Nuxt Devtools...");
    await execa("npm", ["install", "-g", `${name}@latest`], { stdio: "inherit" });
    enable(rc, cwd);
    consola.info("Nuxt Devtools enabled! Restart your Nuxt app to start using it.");
  } else if (command === "disable") {
    const isRemove = args.includes("--remove");
    if (disable(rc, cwd))
      consola.success("Nuxt Devtools disabled for this project.");
    else if (!isRemove)
      consola.warn("Nuxt Devtools is not enabled for this project.");
    if (isRemove) {
      removeModule(rc);
      consola.success("Nuxt Devtools is removed globally");
    }
  } else if (!command) {
    console.log(`npx ${name} enable|disable`);
    process.exit(1);
  } else {
    consola.log(`Unknown command "${command}"`);
    process.exit(1);
  }
  writeUser(rc, RC_PATH);
}
run().catch((err) => {
  consola.error(err);
  process.exit(1);
});
